// @litert/televoke/lib/Server/Errors.js
"use strict";
/**
 * Copyright 2021 Angus.Fenying <fenying@litert.org>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.E_GATEWAY_BUSY = exports.E_SERVER_NOT_READY = exports.E_NO_SERVER_ATTACHED = exports.E_SERVER_BUSY = exports.E_SERVER_CLOSING = exports.E_SERVER_STARTING = exports.E_GATEWAY_CLOSING = exports.E_GATEWAY_STARTING = void 0;
const Errors_1 = require("../Errors");
exports.E_GATEWAY_STARTING = Errors_1.errorRegistry.register({
    name: 'gateway_starting',
    message: 'A gateway has already been starting.',
    metadata: {},
    type: 'public'
});
exports.E_GATEWAY_CLOSING = Errors_1.errorRegistry.register({
    name: 'gateway_closing',
    message: 'A gateway has already been closing.',
    metadata: {},
    type: 'public'
});
exports.E_SERVER_STARTING = Errors_1.errorRegistry.register({
    name: 'server_starting',
    message: 'A server has already been starting.',
    metadata: {},
    type: 'public'
});
exports.E_SERVER_CLOSING = Errors_1.errorRegistry.register({
    name: 'server_closing',
    message: 'A server has already been closing.',
    metadata: {},
    type: 'public'
});
exports.E_SERVER_BUSY = Errors_1.errorRegistry.register({
    name: 'server_busy',
    message: 'It is not able to update server currnetly.',
    metadata: {},
    type: 'public'
});
exports.E_NO_SERVER_ATTACHED = Errors_1.errorRegistry.register({
    name: 'no_server_attached',
    message: 'The gateway does not attach to a server.',
    metadata: {},
    type: 'public'
});
exports.E_SERVER_NOT_READY = Errors_1.errorRegistry.register({
    name: 'server_not_ready',
    message: 'The server is not setup completely.',
    metadata: {},
    type: 'public'
});
exports.E_GATEWAY_BUSY = Errors_1.errorRegistry.register({
    name: 'gateway_busy',
    message: 'The gateway has been attached to a server.',
    metadata: {},
    type: 'public'
});
