// @litert/televoke/lib/Server/SimpleRouter.js
"use strict";
/**
 * Copyright 2021 Angus.Fenying <fenying@litert.org>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSimpleRouter = void 0;
class SimpleRouter {
    constructor() {
        this._handlers = {};
    }
    add(apiName, handler) {
        this._handlers[apiName] = [handler, true];
        return this;
    }
    register(apiName, handler) {
        this._handlers[apiName] = [handler, false];
        return this;
    }
    removeHandler(apiName) {
        delete this._handlers[apiName];
        return this;
    }
    route(name) {
        return this._handlers[name];
    }
    execute(request, descriptor) {
        if (descriptor[1]) {
            return descriptor[0](...request.args);
        }
        else {
            return descriptor[0](request, ...request.args);
        }
    }
    validate(request) {
        return Array.isArray(request.args);
    }
}
function createSimpleRouter() {
    return new SimpleRouter();
}
exports.createSimpleRouter = createSimpleRouter;
