// @litert/televoke/lib/Errors.js
"use strict";
/**
 * Copyright 2021 Angus.Fenying <fenying@litert.org>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.E_INVALID_PACKET = exports.E_PACKET_TOO_LARGE = exports.errorRegistry = void 0;
const $Exceptions = require("@litert/exception");
exports.errorRegistry = $Exceptions.createExceptionRegistry({
    'module': 'televoke.litert.org',
    'types': {
        'public': {
            'index': $Exceptions.createIncreaseCodeIndex(1)
        }
    }
});
exports.E_PACKET_TOO_LARGE = exports.errorRegistry.register({
    name: 'packet_too_large',
    message: 'The packet is larger than the maximum size of packet.',
    metadata: {},
    type: 'public'
});
exports.E_INVALID_PACKET = exports.errorRegistry.register({
    name: 'invalid_packet',
    message: 'The packet is invalid.',
    metadata: {},
    type: 'public'
});
