// @litert/televoke/lib/Client/Errors.js
"use strict";
/**
 * Copyright 2021 Angus.Fenying <fenying@litert.org>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.E_INVALID_RESPONSE = exports.E_SERVER_UNKNOWN_ERROR = exports.E_SERVER_LOGIC_FAILURE = exports.E_API_NOT_FOUND = exports.E_REQUEST_TIMEOUT = exports.E_SERVER_INTERNAL_ERROR = exports.E_CONN_NOT_READY = exports.E_CONN_CLOSING = exports.E_CONN_LOST = exports.E_OPERATION_ABORTED = void 0;
const Errors_1 = require("../Errors");
exports.E_OPERATION_ABORTED = Errors_1.errorRegistry.register({
    name: 'operation_aborted',
    message: 'The CONNECT/CLOSE operation has been cancelled.',
    metadata: {},
    type: 'public'
});
exports.E_CONN_LOST = Errors_1.errorRegistry.register({
    name: 'conn_lost',
    message: 'The connection to remote server lost.',
    metadata: {},
    type: 'public'
});
exports.E_CONN_CLOSING = Errors_1.errorRegistry.register({
    name: 'conn_closing',
    message: 'The connection has been shutting down.',
    metadata: {},
    type: 'public'
});
exports.E_CONN_NOT_READY = Errors_1.errorRegistry.register({
    name: 'conn_not_ready',
    message: 'The connection is not ready yet.',
    metadata: {},
    type: 'public'
});
exports.E_SERVER_INTERNAL_ERROR = Errors_1.errorRegistry.register({
    name: 'server_internal_error',
    message: 'Something wrong insides the remote server.',
    metadata: {},
    type: 'public'
});
exports.E_REQUEST_TIMEOUT = Errors_1.errorRegistry.register({
    name: 'request_timeout',
    message: 'The server does not response in time.',
    metadata: {},
    type: 'public'
});
exports.E_API_NOT_FOUND = Errors_1.errorRegistry.register({
    name: 'api_not_found',
    message: 'The requested API does not exist.',
    metadata: {},
    type: 'public'
});
exports.E_SERVER_LOGIC_FAILURE = Errors_1.errorRegistry.register({
    name: 'server_logic_failure',
    message: 'Failed to handle the request by remote handler.',
    metadata: {},
    type: 'public'
});
exports.E_SERVER_UNKNOWN_ERROR = Errors_1.errorRegistry.register({
    name: 'server_unknown_error',
    message: 'Something unexpected occured.',
    metadata: {},
    type: 'public'
});
exports.E_INVALID_RESPONSE = Errors_1.errorRegistry.register({
    name: 'invalid_response',
    message: 'The response from remote is invalid.',
    metadata: {},
    type: 'public'
});
