// @litert/http-client/lib/Common/Request.js
"use strict";
/**
 * Copyright 2023 Angus.Fenying <fenying@litert.org>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_HTTP_PORT = exports.DEFAULT_HTTPS_PORT = exports.DEFAULT_TIMEOUT = exports.DEFAULT_KEEP_ALIVE_TTL = exports.DEFAULT_PROTOCOL_DETECTION_CACHE_TTL = exports.ETLSVersion = void 0;
var ETLSVersion;
(function (ETLSVersion) {
    ETLSVersion[ETLSVersion["TLS_V1"] = 1] = "TLS_V1";
    ETLSVersion[ETLSVersion["TLS_V1_1"] = 1.1] = "TLS_V1_1";
    ETLSVersion[ETLSVersion["TLS_V1_2"] = 1.2] = "TLS_V1_2";
    ETLSVersion[ETLSVersion["TLS_V1_3"] = 1.3] = "TLS_V1_3";
})(ETLSVersion = exports.ETLSVersion || (exports.ETLSVersion = {}));
exports.DEFAULT_PROTOCOL_DETECTION_CACHE_TTL = 60000;
exports.DEFAULT_KEEP_ALIVE_TTL = 60000;
exports.DEFAULT_TIMEOUT = 30000;
exports.DEFAULT_HTTPS_PORT = 443;
exports.DEFAULT_HTTP_PORT = 80;
