// @litert/http-client/lib/Errors.js
"use strict";
/**
 * Copyright 2023 Angus.Fenying <fenying@litert.org>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.E_NO_RESPONSE_ENTITY = exports.E_TOO_LARGE_RESPONSE_ENTITY = exports.E_PROTOCOL_NOT_SUPPORTED = exports.E_NO_CONTENT_LENGTH = exports.E_UNKNOWN_AUTH_TYPE = exports.E_EMPTY_AUTH_CREDENTIALS = exports.errorRegistry = void 0;
const $Exceptions = require("@litert/exception");
exports.errorRegistry = $Exceptions.createExceptionRegistry({
    'module': 'http-client.litert.org',
    'types': {
        'public': {
            'index': $Exceptions.createIncreaseCodeIndex(1)
        }
    }
});
exports.E_EMPTY_AUTH_CREDENTIALS = exports.errorRegistry.register({
    name: 'empty_auth_credentials',
    message: 'The authentication credential is empty.',
    metadata: {},
    type: 'public'
});
exports.E_UNKNOWN_AUTH_TYPE = exports.errorRegistry.register({
    name: 'unknown_auth_type',
    message: 'The type of authentication is not recognizable.',
    metadata: {},
    type: 'public'
});
exports.E_NO_CONTENT_LENGTH = exports.errorRegistry.register({
    name: 'no_content_length',
    message: 'The header content-length is not specific.',
    metadata: {},
    type: 'public'
});
exports.E_PROTOCOL_NOT_SUPPORTED = exports.errorRegistry.register({
    name: 'protocol_not_supported',
    message: 'The protocol is not supported.',
    metadata: {},
    type: 'public'
});
exports.E_TOO_LARGE_RESPONSE_ENTITY = exports.errorRegistry.register({
    name: 'too_large_response_entity',
    message: 'The entity of response is too large.',
    metadata: {},
    type: 'public'
});
exports.E_NO_RESPONSE_ENTITY = exports.errorRegistry.register({
    name: 'no_response_entity',
    message: 'The response entity is empty.',
    metadata: {},
    type: 'public'
});
