// @litert/typeguard/lib/BuiltInTypes.js
"use strict";
/**
 * Copyright 2022 Angus Fenying <fenying@litert.org>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.UFLOAT = exports.UDECIMAL = exports.FLOAT = exports.NUMBER = exports.DECIMAL = exports.REQUIRED = exports.OPTIONAL = exports.TRUE_VALUE = exports.FALSE_VALUE = exports.TRUE = exports.FALSE = exports.VOID = exports.UNDEFINED = exports.NULL = exports.STRUCT = exports.ANY = exports.ARRAY = exports.BOOLEAN = exports.UINT64 = exports.UINT32 = exports.UINT16 = exports.UINT8 = exports.UINT = exports.SAFE_UINT = exports.INT64 = exports.INT32 = exports.INT16 = exports.INT8 = exports.INT = exports.SAFE_INT = exports.NUMERIC = exports.LATIN_STRING = exports.HEX_STRING = exports.ASCII_STRING = exports.STRING = void 0;
exports.STRING = 'string';
exports.ASCII_STRING = 'ascii_string';
exports.HEX_STRING = 'hex_string';
exports.LATIN_STRING = 'latin_string';
exports.NUMERIC = 'numeric';
exports.SAFE_INT = 'safe_int';
exports.INT = 'int';
exports.INT8 = 'int8';
exports.INT16 = 'int16';
exports.INT32 = 'int32';
exports.INT64 = 'int64';
exports.SAFE_UINT = 'safe_uint';
exports.UINT = 'uint';
exports.UINT8 = 'uint8';
exports.UINT16 = 'uint16';
exports.UINT32 = 'uint32';
exports.UINT64 = 'uint64';
exports.BOOLEAN = 'boolean';
exports.ARRAY = 'array';
exports.ANY = 'any';
exports.STRUCT = 'struct';
exports.NULL = 'null';
exports.UNDEFINED = 'undefined';
exports.VOID = 'void';
exports.FALSE = 'false';
exports.TRUE = 'true';
exports.FALSE_VALUE = 'false_value';
exports.TRUE_VALUE = 'true_value';
exports.OPTIONAL = 'optional';
exports.REQUIRED = 'required';
exports.DECIMAL = 'decimal';
exports.NUMBER = 'number';
exports.FLOAT = 'float';
exports.UDECIMAL = 'udecimal';
exports.UFLOAT = 'ufloat';
