// @litert/typeguard/lib/Internal.js
"use strict";
/**
 * Copyright 2022 Angus Fenying <fenying@litert.org>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RE_VALID_CUSTOM_TYPE_NAME = exports.EFlagValue = exports.EFlags = exports.KEY_EQUAL_SUFFIX = exports.KEY_STRICT_SUFFIX = exports.KEY_ARRAY_SUFFIX = exports.KEY_LIST_SUFFIX = exports.KEY_MAP_SUFFIX = exports.NEGATIVE_SYMBOL = exports.PREDEF_TYPE_SYMBOL = exports.IMPLICIT_SYMBOL = exports.FILTER_PREFIX = exports.MODIFIER_PREFIX = exports.LIST_SUFFIX = exports.MAP_SUFFIX = void 0;
exports.MAP_SUFFIX = '{}';
exports.LIST_SUFFIX = '[]';
exports.MODIFIER_PREFIX = '$.';
exports.FILTER_PREFIX = '|';
exports.IMPLICIT_SYMBOL = '?';
exports.PREDEF_TYPE_SYMBOL = '@';
exports.NEGATIVE_SYMBOL = '!';
exports.KEY_MAP_SUFFIX = `->${exports.MAP_SUFFIX}`;
exports.KEY_LIST_SUFFIX = `->${exports.LIST_SUFFIX}`;
exports.KEY_ARRAY_SUFFIX = /->\[\s*(\d+)(\s*,\s*(\d+)?)?\s*\]$/;
exports.KEY_STRICT_SUFFIX = '->()';
exports.KEY_EQUAL_SUFFIX = '->(=)';
var EFlags;
(function (EFlags) {
    EFlags[EFlags["FROM_STRING"] = 0] = "FROM_STRING";
    EFlags[EFlags["STRICT"] = 1] = "STRICT";
    EFlags[EFlags["OPTIONAL"] = 2] = "OPTIONAL";
    EFlags[EFlags["REQUIRED"] = 3] = "REQUIRED";
    EFlags[EFlags["ARRAY"] = 4] = "ARRAY";
})(EFlags = exports.EFlags || (exports.EFlags = {}));
var EFlagValue;
(function (EFlagValue) {
    /**
     * Disabled.
     */
    EFlagValue[EFlagValue["NO"] = 0] = "NO";
    /**
     * Enabled but not inheritable.
     */
    EFlagValue[EFlagValue["YES"] = 1] = "YES";
    /**
     * Enabled and inheritable if not deep into sub element.
     */
    EFlagValue[EFlagValue["INHERIT"] = 2] = "INHERIT";
    /**
     * Enabled and inheritable even deep into sub element.
     */
    EFlagValue[EFlagValue["ELEMENT_INHERIT"] = 3] = "ELEMENT_INHERIT";
})(EFlagValue = exports.EFlagValue || (exports.EFlagValue = {}));
exports.RE_VALID_CUSTOM_TYPE_NAME = /^[-:.\w]+$/;
