// @litert/typeguard/lib/Modifiers.js
"use strict";
/**
 * Copyright 2022 Angus Fenying <fenying@litert.org>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TYPE = exports.TAG = exports.EQUAL = exports.TUPLE = exports.STRING = exports.STRICT = exports.DICT = exports.ARRAY = exports.LIST = exports.MAP = exports.AND = exports.OR = exports.NOT = exports.PREFIX = void 0;
const I = require("./Internal");
exports.PREFIX = I.MODIFIER_PREFIX;
exports.NOT = `${I.MODIFIER_PREFIX}not`;
exports.OR = `${I.MODIFIER_PREFIX}or`;
exports.AND = `${I.MODIFIER_PREFIX}and`;
exports.MAP = `${I.MODIFIER_PREFIX}map`;
exports.LIST = `${I.MODIFIER_PREFIX}list`;
exports.ARRAY = `${I.MODIFIER_PREFIX}array`;
exports.DICT = `${I.MODIFIER_PREFIX}dict`;
exports.STRICT = `${I.MODIFIER_PREFIX}strict`;
exports.STRING = `${I.MODIFIER_PREFIX}string`;
exports.TUPLE = `${I.MODIFIER_PREFIX}tuple`;
exports.EQUAL = `${I.MODIFIER_PREFIX}equal`;
exports.TAG = `${I.MODIFIER_PREFIX}tag`;
exports.TYPE = `${I.MODIFIER_PREFIX}type`;
