// @litert/televoke/lib/Common/Response.js
"use strict";
/**
 * Copyright 2021 Angus.Fenying <fenying@litert.org>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EResponseCode = void 0;
var EResponseCode;
(function (EResponseCode) {
    /**
     * API was called succeed
     */
    EResponseCode[EResponseCode["OK"] = 0] = "OK";
    /**
     * The invoked API does not exist on the remote server.
     */
    EResponseCode[EResponseCode["API_NOT_FOUND"] = 1] = "API_NOT_FOUND";
    /**
     * Some unexpected error happened.
     */
    EResponseCode[EResponseCode["SYSTEM_ERROR"] = 2] = "SYSTEM_ERROR";
    /**
     * Logical error happened inside the API handler.
     */
    EResponseCode[EResponseCode["FAILURE"] = 3] = "FAILURE";
    /**
     * The arguments inside the request is malformed.
     */
    EResponseCode[EResponseCode["MALFORMED_ARGUMENTS"] = 4] = "MALFORMED_ARGUMENTS";
})(EResponseCode = exports.EResponseCode || (exports.EResponseCode = {}));
