// @litert/clap/lib/Errors.js
"use strict";
/**
 *  Copyright 2021 Angus.Fenying <fenying@litert.org>
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.E_ARGUMENTS_LACKED = exports.E_COMMAND_REQUIRED = exports.E_OPTION_VALUE_REQUIRED = exports.E_INVALID_FLAG_SHORTCUT = exports.E_INVALID_COMMAND_SHORTCUT = exports.E_INVALID_OPTION_SHORTCUT = exports.E_INVALID_FLAG_NAME = exports.E_INVALID_COMMAND_NAME = exports.E_INVALID_OPTION_NAME = exports.E_DUP_FLAG_SHORTCUT = exports.E_DUP_COMMAND_SHORTCUT = exports.E_DUP_OPTION_SHORTCUT = exports.E_DUP_FLAG_NAME = exports.E_DUP_COMMAND_NAME = exports.E_DUP_OPTION_NAME = exports.E_NO_SUCH_FLAG = exports.E_NO_SUCH_OPTION = exports.E_NO_SUCH_COMMAND = exports.errorRegistry = void 0;
const $Exceptions = require("@litert/exception");
exports.errorRegistry = $Exceptions.createExceptionRegistry({
    'module': 'clap.litert.org',
    'types': {
        'prepare': {
            'index': $Exceptions.createIncreaseCodeIndex(1)
        },
        'runtime': {
            'index': $Exceptions.createIncreaseCodeIndex(1000)
        }
    }
});
exports.E_NO_SUCH_COMMAND = exports.errorRegistry.register({
    name: 'no_such_command',
    message: 'The command does not exist.',
    metadata: {},
    type: 'prepare'
});
exports.E_NO_SUCH_OPTION = exports.errorRegistry.register({
    name: 'no_such_option',
    message: 'The option does not exist.',
    metadata: {},
    type: 'prepare'
});
exports.E_NO_SUCH_FLAG = exports.errorRegistry.register({
    name: 'no_such_flag',
    message: 'The flag does not exist.',
    metadata: {},
    type: 'prepare'
});
exports.E_DUP_OPTION_NAME = exports.errorRegistry.register({
    name: 'dup_option_name',
    message: 'The name of option has already been used.',
    metadata: {},
    type: 'prepare'
});
exports.E_DUP_COMMAND_NAME = exports.errorRegistry.register({
    name: 'dup_command_name',
    message: 'The name of command has already been used.',
    metadata: {},
    type: 'prepare'
});
exports.E_DUP_FLAG_NAME = exports.errorRegistry.register({
    name: 'dup_flag_name',
    message: 'The name of flag has already been used.',
    metadata: {},
    type: 'prepare'
});
exports.E_DUP_OPTION_SHORTCUT = exports.errorRegistry.register({
    name: 'dup_option_shortcut',
    message: 'The shortcut of option has already been used.',
    metadata: {},
    type: 'prepare'
});
exports.E_DUP_COMMAND_SHORTCUT = exports.errorRegistry.register({
    name: 'dup_command_shortcut',
    message: 'The shortcut of command has already been used.',
    metadata: {},
    type: 'prepare'
});
exports.E_DUP_FLAG_SHORTCUT = exports.errorRegistry.register({
    name: 'dup_flag_shortcut',
    message: 'The shortcut of flag has already been used.',
    metadata: {},
    type: 'prepare'
});
exports.E_INVALID_OPTION_NAME = exports.errorRegistry.register({
    name: 'invalid_option_name',
    message: 'The name of option is malformed.',
    metadata: {},
    type: 'prepare'
});
exports.E_INVALID_COMMAND_NAME = exports.errorRegistry.register({
    name: 'invalid_command_name',
    message: 'The name of command is malformed.',
    metadata: {},
    type: 'prepare'
});
exports.E_INVALID_FLAG_NAME = exports.errorRegistry.register({
    name: 'invalid_flag_name',
    message: 'The name of flag is malformed.',
    metadata: {},
    type: 'prepare'
});
exports.E_INVALID_OPTION_SHORTCUT = exports.errorRegistry.register({
    name: 'invalid_option_shortcut',
    message: 'The shortcut of option is malformed.',
    metadata: {},
    type: 'prepare'
});
exports.E_INVALID_COMMAND_SHORTCUT = exports.errorRegistry.register({
    name: 'invalid_command_shortcut',
    message: 'The shortcut of command is malformed.',
    metadata: {},
    type: 'prepare'
});
exports.E_INVALID_FLAG_SHORTCUT = exports.errorRegistry.register({
    name: 'invalid_flag_shortcut',
    message: 'The shortcut of flag is malformed.',
    metadata: {},
    type: 'prepare'
});
exports.E_OPTION_VALUE_REQUIRED = exports.errorRegistry.register({
    name: 'option_value_required',
    message: 'The value of option is required.',
    metadata: {},
    type: 'runtime'
});
exports.E_COMMAND_REQUIRED = exports.errorRegistry.register({
    name: 'command_required',
    message: 'The command is required before arguments.',
    metadata: {},
    type: 'runtime'
});
exports.E_ARGUMENTS_LACKED = exports.errorRegistry.register({
    name: 'arguments_lacked',
    message: 'Too few arguments.',
    metadata: {},
    type: 'runtime'
});
