// @litert/core/dist/class.Exception.js
"use strict";
/*
   +----------------------------------------------------------------------+
   | LiteRT Core.js Library                                               |
   +----------------------------------------------------------------------+
   | Copyright (c) 2018 Fenying Studio                                    |
   +----------------------------------------------------------------------+
   | This source file is subject to version 2.0 of the Apache license,    |
   | that is bundled with this package in the file LICENSE, and is        |
   | available through the world-wide-web at the following url:           |
   | https://github.com/litert/core.js/blob/master/LICENSE                |
   +----------------------------------------------------------------------+
   | Authors: Angus Fenying <fenying@litert.org>                          |
   +----------------------------------------------------------------------+
 */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * This class provides the default exception class for LiteRT.
 */
class Exception {
    /**
     * The data needs to be carried by exception.
     */
    get origin() {
        return this._origin;
    }
    /**
     * The trace of Exception.
     *
     * This is an array contains the calling stacks from the new operation of
     * an exception.
     */
    get trace() {
        return this._trace;
    }
    /**
     * The type of the exception.
     */
    get type() {
        return this._type;
    }
    /**
     * The code to identify the type of this exception.
     */
    get error() {
        return this._errno;
    }
    /**
     * The description about this exception.
     */
    get message() {
        return this._message;
    }
    /**
     * Constructor of an exception.
     *
     * @param error The code to identify the type of this exception.
     * @param message The description about this exception.
     * @param origin The data needs to be carried by exception.
     */
    constructor(error, message, origin) {
        this._errno = error;
        this._message = message;
        this._origin = origin;
        this._trace = new Error().stack.split(/\s+at\s+/).slice(3);
        /**
         * this._type should be defined in the sub classes.
         */
    }
    /**
     * @override
     */
    toString() {
        return `Error(${this._type} ${this._errno}): ${this.message}`;
    }
    /**
     * Convert this exception into JSON string.
     */
    toJSON() {
        return JSON.stringify({
            "type": this._type,
            "error": this._errno,
            "message": this._message,
            "origin": this._origin,
            "trace": this._trace
        });
    }
    /**
     * THe method helps determine if an error object is of LiteRT's exception.
     *
     * @param e     The error object to be check.
     * @param type  The type of exception to be ensured.
     */
    static is(e, type) {
        return e && (e instanceof Exception || (typeof e._type === "string" &&
            typeof e._error === "number" &&
            typeof e._message === "string" &&
            typeof e._trace === "object")) && (type === void 0 || type === e.type);
    }
}
exports.Exception = Exception;
