// @litert/http/index.js
"use strict";
/*
   +----------------------------------------------------------------------+
   | LiteRT HTTP.js Library                                               |
   +----------------------------------------------------------------------+
   | Copyright (c) 2018 Fenying Studio                                    |
   +----------------------------------------------------------------------+
   | This source file is subject to version 2.0 of the Apache license,    |
   | that is bundled with this package in the file LICENSE, and is        |
   | available through the world-wide-web at the following url:           |
   | https://github.com/litert/http.js/blob/master/LICENSE                |
   +----------------------------------------------------------------------+
   | Authors: Angus Fenying <fenying@litert.org>                          |
   +----------------------------------------------------------------------+
 */
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var Abstract_1 = require("./server/Abstract");
exports.ServerStatus = Abstract_1.ServerStatus;
exports.DEFAULT_BACKLOG = Abstract_1.DEFAULT_BACKLOG;
exports.DEFAULT_HOST = Abstract_1.DEFAULT_HOST;
exports.DEFAULT_PORT = Abstract_1.DEFAULT_PORT;
exports.DEFAULT_SSL_PORT = Abstract_1.DEFAULT_SSL_PORT;
exports.DEFAULT_EXPECT_REQUEST = Abstract_1.DEFAULT_EXPECT_REQUEST;
exports.DEFAULT_KEEP_ALIVE = Abstract_1.DEFAULT_KEEP_ALIVE;
exports.EXCEPTION_TYPE = Abstract_1.EXCEPTION_TYPE;
exports.HTTP_METHODS = Abstract_1.HTTP_METHODS;
exports.HTTPStatus = Abstract_1.HTTPStatus;
exports.CookiesEncoding = Abstract_1.CookiesEncoding;
const plugins = require("./server/built-in-plugins");
exports.plugins = plugins;
var built_in_plugins_1 = require("./server/built-in-plugins");
/**
 * @deprecated Use method `createCookiesEncoder` instead, this method will
 *             be removed in v0.5.0.
 */
exports.createStandardCookiesEncoder = built_in_plugins_1.createCookiesEncoder;
const MountableHost_1 = require("./server/MountableHost");
const StandardDispatcher_1 = require("./server/StandardDispatcher");
const StandardHost_1 = require("./server/StandardHost");
var StandardRouter_1 = require("./server/StandardRouter");
exports.createStandardRouter = StandardRouter_1.createStandardRouter;
__export(require("./server/ControllerRouter"));
const Errors_1 = require("./server/Errors");
exports.ServerError = Errors_1.default;
const Exception_1 = require("./server/Exception");
exports.Exception = Exception_1.default;
function createServer(opts) {
    return new StandardHost_1.default(opts);
}
exports.createServer = createServer;
function createMountableServer(opts) {
    return new MountableHost_1.default(opts);
}
exports.createMountableServer = createMountableServer;
function createVirtualDispatcher(opts) {
    return new StandardDispatcher_1.default(opts);
}
exports.createVirtualDispatcher = createVirtualDispatcher;
var DefaultContext_1 = require("./server/DefaultContext");
exports.createDefaultContext = DefaultContext_1.createDefaultContext;
exports.DefaultContext = DefaultContext_1.DefaultContext;
