// @litert/exception/examples/quick-start.js
"use strict";
/**
 * Copyright 2022 Angus.Fenying <fenying@litert.org>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable @typescript-eslint/no-magic-numbers */
const $Exceptions = require("../lib");
const registry = $Exceptions.createExceptionRegistry({
    module: 'exception.litert.org',
    types: {
        'private': {
            index: $Exceptions.createDecreaseCodeIndex(-1)
        },
        'public': {
            index: $Exceptions.createIncreaseCodeIndex(1)
        }
    }
});
const registry2 = $Exceptions.createExceptionRegistry({
    module: 'exception2.litert.org',
    types: {
        'private': {
            index: $Exceptions.createDecreaseCodeIndex(-1)
        },
        'public': {
            index: $Exceptions.createIncreaseCodeIndex(1)
        }
    }
});
const E_NO_USER = registry.register({
    type: 'public',
    name: 'no_user',
    message: 'No such a user.',
    metadata: {}
});
const E_SYNTAX_ERROR = registry.register({
    type: 'private',
    name: 'syntax_error',
    message: 'Unexpected syntax error found.',
    metadata: {}
});
const synErr = new E_SYNTAX_ERROR();
const noUserErr = new E_NO_USER();
console.error('Exception URI:      ', synErr.toString());
console.error('Exception URI:      ', noUserErr.toString(true));
console.info('Exception name:     ', (_a = registry.parse(noUserErr.toString(true))) === null || _a === void 0 ? void 0 : _a.name);
console.info('Exception message:  ', (_b = registry.parse(noUserErr.toString(true))) === null || _b === void 0 ? void 0 : _b.message);
console.info('Exception module:   ', (_c = registry.parse(noUserErr.toString(true))) === null || _c === void 0 ? void 0 : _c.module);
console.info('Exception code:     ', (_d = registry.parse(noUserErr.toString(true))) === null || _d === void 0 ? void 0 : _d.code);
console.info('Exception name:     ', registry.fromJSON(noUserErr.toJSON(true)).name);
console.info('Exception message:  ', registry.fromJSON(noUserErr.toJSON(true)).message);
console.info('Exception module:   ', registry.fromJSON(noUserErr.toJSON(true)).module);
console.info('Exception code:     ', registry.fromJSON(noUserErr.toJSON(true)).code);
console.info('Exception stack:    ', registry.fromJSON(noUserErr.toJSON(true)).stack);
console.info('Exception stack:    ', registry.fromJSON(noUserErr.toJSON()).stack);
console.info('Exception name:     ', registry2.fromJSON(noUserErr.toJSON(true)).name);
console.info('Exception message:  ', registry2.fromJSON(noUserErr.toJSON(true)).message);
console.info('Exception module:   ', registry2.fromJSON(noUserErr.toJSON(true)).module);
console.info('Exception code:     ', registry2.fromJSON(noUserErr.toJSON(true)).code);
console.info('Exception stack:    ', registry2.fromJSON(noUserErr.toJSON(true)).stack);
console.info('Exception stack:    ', registry2.fromJSON(noUserErr.toJSON()).stack);
console.info('Exception message:  ', noUserErr.message);
console.error('Exception stack:    ', (_e = registry.parse(noUserErr.toString())) === null || _e === void 0 ? void 0 : _e.stack);
if (registry.identify(noUserErr, 'public', 'no_user')) {
    console.info('It is a NO_USER exception.');
}
try {
    registry.register({
        type: '-public',
        name: 'no_user',
        message: 'No such a user.',
        metadata: {}
    });
}
catch (e) {
    if ($Exceptions.identify(e, $Exceptions.E_TYPE_NOT_FOUND.type, $Exceptions.E_TYPE_NOT_FOUND.name)) {
        console.info('MATCHED(identify): E_TYPE_NOT_FOUND');
    }
    if ($Exceptions.equals(e, $Exceptions.E_TYPE_NOT_FOUND)) {
        console.info('MATCHED(equals): E_TYPE_NOT_FOUND');
    }
    console.error('Exception URI:    ', e.toString());
}
if (registry.has('no_user')) {
    console.info('Exception "no_user" has been defined.');
    registry2.register({
        type: 'public',
        name: 'no_user',
        message: 'No such a user.',
        metadata: {}
    });
    console.info('Successfully defined the exception "no_user" in another registry.');
}
console.info(JSON.stringify(registry.getDefinitions(), null, 2));
registry.register({
    type: 'public',
    name: 'no_user',
    message: 'No such a user.',
    metadata: {}
});
